packages/nx/src/project-graph/plugins/internal-api.ts

```diff
@@ -97,3 +113,73 @@ export async function getDefaultPlugins(root: string) {
     join(__dirname, '../../plugins/project-json/build-nodes/project-json'),
   ];
 }
+
+let projectsWithoutInference: Record<string, ProjectConfiguration>;
+
+export async function loadPlugins(
+  plugins: PluginConfiguration[],
+  root: string
+): Promise<LoadedNxPlugin[]> {
+  return await Promise.all(plugins.map((p) => loadPlugin(p, root)));
+}
+
+export async function loadPlugin(plugin: PluginConfiguration, root: string) {
+  try {
+    require.resolve(typeof plugin === 'string' ? plugin : plugin.plugin);
+  } catch {
+    // If a plugin cannot be resolved, we will need projects to resolve it
+    projectsWithoutInference ??=
+      await retrieveProjectConfigurationsWithoutPluginInference(root);
+  }
+  return await loadNxPluginAsync(
+    plugin,
+    getNxRequirePaths(root),
+    projectsWithoutInference,
+    root
+  );
+}
+
+export type LoadedNxPlugin = {
+  plugin: NxPlugin;
+  options?: unknown;
+};
+
+export async function loadNxPluginAsync(
+  pluginConfiguration: PluginConfiguration,
+  paths: string[],
+  projects: Record<string, ProjectConfiguration>,
+  root: string
+): Promise<LoadedNxPlugin> {
+  const { plugin: moduleName, options } =
+    typeof pluginConfiguration === 'object'
+      ? pluginConfiguration
+      : { plugin: pluginConfiguration, options: undefined };
+
+  performance.mark(`Load Nx Plugin: ${moduleName} - start`);
+  let { pluginPath, name } = await getPluginPathAndName(
+    moduleName,
+    paths,
+    projects,
+    root
+  );
+  const plugin = normalizeNxPlugin(await importPluginModule(pluginPath));
+  plugin.name ??= name;
+  performance.mark(`Load Nx Plugin: ${moduleName} - end`);
+  performance.measure(
+    `Load Nx Plugin: ${moduleName}`,
+    `Load Nx Plugin: ${moduleName} - start`,
+    `Load Nx Plugin: ${moduleName} - end`
+  );
+  return { plugin, options };
+}
```

Comment:
I'm really not a big fan of this making it over into the `internal-api` file. Ideally, The plugin worker should only have to import from worker-api and utils, to ensure that there is a clean separation between the stuff happening there and the rest of the code.

Additionally, if a plugin were to set its create nodes pattern based on the result of a config file (similar to how the package-json-workspaces plugin works) we could very easily hit a situation where swc-node and ts-node are both registered on main thread. Imagine if instead of package.json having the workspaces field it was something like workspace.ts that required a transpiler.

If the only reason we are doing this is to support listing `createNodes` and `createDependencies` support in `nx list`, I'd honestly rather drop that support and we could add some field to package.json if a plugin wants to list that they support it or similar. e.g. something like

```json
{
  "nx-plugin": {
    "generators": "generators.json"
    "executors": "executors.json"
    "graph": ... // this could be a boolean to indicate support, but maybe a path to import would work well here. Then we would be able to drop the /plugin at the end of a lot of our PCV3 plugins 🤔 .
  }
}
```